src/domain-services/flows/flow-service.ts

```diff
@@ -0,0 +1,544 @@
+import { type Database } from '@unocha/hpc-api-core/src/db';
+import { type FlowId } from '@unocha/hpc-api-core/src/db/models/flow';
+import { Op } from '@unocha/hpc-api-core/src/db/util/conditions';
+import { type InstanceOfModel } from '@unocha/hpc-api-core/src/db/util/types';
+import { createBrandedValue } from '@unocha/hpc-api-core/src/util/types';
+import { Service } from 'typedi';
+import { FlowObjectService } from '../flow-object/flow-object-service';
+import type {
+  FlowObject,
+  FlowObjectFilterGrouped,
+  FlowObjectType,
+} from '../flow-object/model';
+import { buildWhereConditionsForFlowObjectFilters } from '../flow-object/utils';
+import { type FlowParkedParentSource } from './graphql/types';
+import type {
+  FlowInstance,
+  FlowOrderBy,
+  FlowOrderByCond,
+  FlowWhere,
+  IGetFlowsArgs,
+  UniqueFlowEntity,
+} from './model';
+import { buildSearchFlowsConditions } from './strategy/impl/utils';
+
+@Service()
+export class FlowService {
+  constructor(private readonly flowObjectService: FlowObjectService) {}
+
+  async getFlows(args: IGetFlowsArgs): Promise<FlowInstance[]> {
+    const { models, orderBy, conditions, limit, offset } = args;
+
+    const distinctColumns = [
+      'id' as keyof FlowInstance,
+      'versionID' as keyof FlowInstance,
+    ];
+
+    if (orderBy) {
+      distinctColumns.push(orderBy.column as keyof FlowInstance);
+      distinctColumns.reverse();
+    }
+
+    const flows: FlowInstance[] = await models.flow.find({
+      orderBy,
+      where: conditions,
+      distinct: distinctColumns,
+      limit,
+      offset,
+    });
+
+    return flows;
+  }
+
+  async getFlowIDsFromEntity(
+    database: Database,
+    orderBy: FlowOrderBy
+  ): Promise<UniqueFlowEntity[]> {
+    const entity = orderBy.subEntity ?? orderBy.entity;
+    // Get the entity list
+    // 'externalReference' is a special case
+    // because it does have a direct relation with flow
+    // and no direction
+    if (entity === 'externalReference') {
+      const column = orderBy.column as keyof InstanceOfModel<
+        Database['externalReference']
+      >;
+      const externalReferences = await database.externalReference.find({
+        orderBy: { column, order: orderBy.order },
+        distinct: ['flowID', 'versionID'],
+      });
+
+      const uniqueFlowEntities: UniqueFlowEntity[] = externalReferences.map(
+        (externalReference) =>
+          ({
+            id: externalReference.flowID,
+            versionID: externalReference.versionID,
+          }) satisfies UniqueFlowEntity
+      );
+
+      return uniqueFlowEntities;
+    }
+
+    const refDirection = orderBy.direction ?? 'source';
+
+    // Validate the variable using io-ts
+
+    let flowObjects = [];
+    let entityIDsSorted: number[] = [];
+
+    switch (entity) {
+      case 'emergency': {
+        // Get emergency entities sorted
+        const column = orderBy.column as keyof InstanceOfModel<
```

Comment:
We can use intersection type to handle this kind of issues. For example.
```ts
type FlowOrderByCommon = {
  order: SortOrder;
  direction?: EntityDirection;
};

export type FlowOrderBy = FlowOrderByCommon &
  (
    | {
        column: keyof FlowInstance;
        entity: 'flow';
      }
    | {
        column: keyof InstanceOfModel<Database['externalReference']>;
        entity: 'externalReference';
      }
    | {
        column: keyof InstanceOfModel<Database['emergency']>;
        entity: 'emergency';
      }
    | {
        entity: 'globalCluster';
        column: keyof InstanceOfModel<Database['globalCluster']>;
      }
    | {
        entity: 'governingEntity';
        column: keyof InstanceOfModel<Database['governingEntity']>;
      }
    | {
        entity: 'location';
        column: keyof InstanceOfModel<Database['location']>;
      }
    | {
        entity: 'organization';
        column: keyof InstanceOfModel<Database['organization']>;
      }
    | {
        entity: 'plan';
        column: keyof InstanceOfModel<Database['plan']>;
      }
    | {
        entity: 'usageYear';
        column: keyof InstanceOfModel<Database['usageYear']>;
      }
    | {
        entity: 'planVersion';
        column: keyof InstanceOfModel<Database['planVersion']>;
      }
    | {
        entity: 'project';
        column: keyof InstanceOfModel<Database['project']>;
      }
  );
```
We can also make it `io-ts` codec so we will have type checking in runtime. We can check if it is correct type or not on function https://github.com/UN-OCHA/hpc-api/blob/c6f882510e0154e65367fbdd4d0a2796239f586a/src/domain-services/flows/strategy/impl/utils.ts#L311

Reply:
Did you miss the last part of the comment? It looks like 😅